### ISSUE TITLE
Error Loading Weights with `by_name=True`

### ISSUE THREAD

**Issue by hamelsmu** [tensorflow/tensorflow#38251]
Here is a short Colab notebook with this error

https://colab.research.google.com/drive/1Z9RoOtyPw-MXnDgPK5irbuZ9I6tpz-zM

**Describe the current behavior**
Error message complains that I must have `TensorFlow-formatted weights` however when I am saving weights I am saving this as TensorFlow formatted weights, using the below code:

```py
model.save_weights('model_weights', save_format='tf')`
```

**Describe the expected behavior**

**Standalone code to reproduce the issue** 

https://colab.research.google.com/drive/1Z9RoOtyPw-MXnDgPK5irbuZ9I6tpz-zM

**Comment by aniketmaurya** [comment]
@hamelsmu  While loading your weights set **by_name=False** :
`new_model.load_weights('model_weights', by_name=False)`

**Comment by aniketmaurya** [comment]
While loading your weights set **by_name=False** :
`new_model.load_weights('model_weights', by_name=False)`

**Comment by hamelsmu** [comment]
@aniketmaurya  what if I want to load the weights by name?  I want to make sure that I only load weights where the layer names match.  Thanks for your help!

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.2.0rc2](https://colab.research.google.com/gist/amahendrakar/b13ea01f9d05c2d4150b00d8255697ce/38251-2-2.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/7a6e5b16aaacf0d457acf62559df8879/38251-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by jvishnuvardhan** [comment]
@hamelsmu 
TL/DR: So if you want to load a model saved in *.tf format, then select `by_name=False`. If you want to load a model saved in *.h5 format, then select `by_name=True`. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/c0c9808471249fb63298d9273b827f1e/38251-tf-nightly.ipynb).

There is small difference between loading a saved model when the model was saved in *.tf format (tensorflow format) or in *.h5 format. The following error clearly shows what we need to do when we load a tensorflow-formatted weights (it need by_name =False). 

`NotImplementedError: Weights may only be loaded based on topology into Models when loading TensorFlow-formatted weights (got by_name=True to load_weights).`

Check the description of `load_weights` on [TF website](https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly#load_weights). 

`Only topological loading (by_name=False) is supported when loading weights from the TensorFlow format. Note that topological loading differs slightly between TensorFlow and HDF5 formats for user-defined classes inheriting from tf.keras.Model: HDF5 loads based on a flattened list of weights, while the TensorFlow format loads based on the object-local names of attributes to which layers are assigned in the Model's constructor. `

Hope this helps. Thanks! 

Please close the issue if this was resolved for you. Thanks!

**Comment by hamelsmu** [comment]
@jvishnuvardhan I am loading weights from the TF format, so shouldn't `by_name=True` work?

**Comment by hamelsmu** [comment]
Closing this issue, thanks for the clarification.  I clearly missed the error message.

Thank you so much for your help.
